File: recipes/subms-ts-sql/java/src/main/java/com/submillisecond/recipes/tssql/Ast.java
package com.submillisecond.recipes.tssql;

import java.util.List;
import java.util.Optional;

/**
 * The parsed-statement IR. {@link TsSqlStmt} is the whole SELECT after the
 * parser has validated structure but before any lowering decision; the lowerer
 * in {@link Lower} walks it and decides whether the query is a grouped aggregate
 * (lower to {@code subms-ts-groupby}) or a row-wise projection pipeline (lower
 * to {@code subms-ts-lazy}). The nested types mirror the Rust sibling's
 * {@code ast} module.
 */
final class Ast {

    private Ast() {}

    /** A scalar literal as written in the SQL text. */
    sealed interface SqlLiteral permits IntLit, NumLit, StrLit {}

    /** An integer literal with no decimal point or exponent. */
    record IntLit(long value) implements SqlLiteral {}

    /** A floating literal. */
    record NumLit(double value) implements SqlLiteral {}

    /** A single-quoted string literal. */
    record StrLit(String value) implements SqlLiteral {}

    /** Arithmetic operators in a scalar expression. */
    enum ArithOp {
        ADD,
        SUB,
        MUL,
        DIV
    }

    /** Comparison operators in a predicate. */
    enum CmpOp {
        EQ,
        NE,
        LT,
        LE,
        GT,
        GE
    }

    /** The aggregate functions the subset supports. */
    enum AggFunc {
        SUM,
        AVG,
        MIN,
        MAX,
        COUNT
    }

    /**
     * A scalar / boolean expression. The same tree covers a projection scalar, a
     * WHERE predicate, and an aggregate operand; the lowerer enforces which
     * shapes are legal where. Comparison + boolean nodes only ever appear inside
     * a predicate.
     */
    sealed interface SqlExpr
            permits Column, Literal, Arith, Compare, And, Or, Not, Case, Aggregate {

        /**
         * Does this expression contain an aggregate call anywhere in its tree?
         * Drives the grouped-vs-rowwise lowering decision.
         */
        default boolean containsAggregate() {
            return switch (this) {
                case Aggregate a -> true;
                case Column c -> false;
                case Literal l -> false;
                case Arith a -> a.lhs().containsAggregate() || a.rhs().containsAggregate();
                case Compare c -> c.lhs().containsAggregate() || c.rhs().containsAggregate();
                case And n -> n.lhs().containsAggregate() || n.rhs().containsAggregate();
                case Or n -> n.lhs().containsAggregate() || n.rhs().containsAggregate();
                case Not n -> n.operand().containsAggregate();
                case Case k -> k.when().containsAggregate()
                        || k.then().containsAggregate()
                        || k.otherwise().containsAggregate();
            };
        }
    }

    /** A column reference. */
    record Column(String name) implements SqlExpr {}

    /** A literal scalar. */
    record Literal(SqlLiteral value) implements SqlExpr {}

    /** Binary arithmetic over two scalar sub-expressions. */
    record Arith(ArithOp op, SqlExpr lhs, SqlExpr rhs) implements SqlExpr {}

    /** A comparison; yields a boolean. Legal only inside a predicate. */
    record Compare(CmpOp op, SqlExpr lhs, SqlExpr rhs) implements SqlExpr {}

    /** Boolean AND of two predicates. */
    record And(SqlExpr lhs, SqlExpr rhs) implements SqlExpr {}

    /** Boolean OR of two predicates. */
    record Or(SqlExpr lhs, SqlExpr rhs) implements SqlExpr {}

    /** Boolean NOT of a predicate. */
    record Not(SqlExpr operand) implements SqlExpr {}

    /** {@code CASE WHEN <pred> THEN <e> ELSE <e> END}. Lowers to a When. */
    record Case(SqlExpr when, SqlExpr then, SqlExpr otherwise) implements SqlExpr {}

    /** An aggregate call. {@code arg} is empty for {@code COUNT(*)}. */
    record Aggregate(AggFunc func, Optional<SqlExpr> arg) implements SqlExpr {}

    /** One item in the SELECT projection list. */
    sealed interface SelectItem permits Star, ExprItem {}

    /** {@code *} - every source column, in source order. */
    record Star() implements SelectItem {}

    /** A projected expression with its output column name. */
    record ExprItem(SqlExpr expr, String alias) implements SelectItem {}

    /** One ORDER BY key. */
    record OrderKey(String column, boolean ascending) {}

    /** A parsed SELECT statement. Absent clauses are empty / {@code null}. */
    record TsSqlStmt(
            List<SelectItem> projection,
            String table,
            SqlExpr filter,
            List<String> groupBy,
            List<OrderKey> orderBy,
            Optional<Integer> limit) {}
}

File: recipes/subms-ts-sql/rust/src/ast.rs
//! The parsed-statement IR. A [`TsSqlStmt`] is the whole `SELECT` after the
//! parser has validated structure but before any lowering decision; the
//! lowerer in [`crate::lower`] walks it and decides whether the query is a
//! grouped aggregate (lower to `subms-ts-groupby`) or a row-wise projection
//! pipeline (lower to `subms-ts-lazy`). The variants are public so a caller
//! can inspect a parse without executing - `parse(sql)` returns this directly.

/// A scalar literal as written in the SQL text. The lowerer maps these onto
/// `subms-ts-expr` typed literals (`Num` -> f64, `Int` -> i64, `Str` -> string).
#[derive(Clone, Debug, PartialEq)]
pub enum SqlLiteral {
    /// An integer literal with no decimal point or exponent (`42`).
    Int(i64),
    /// A floating literal (`3.5`, `1e3`).
    Num(f64),
    /// A single-quoted string literal (`'AAPL'`).
    Str(String),
}

/// Arithmetic operators in a scalar expression.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum ArithOp {
    Add,
    Sub,
    Mul,
    Div,
}

/// Comparison operators in a predicate. `=`/`<>` map onto `TsExpr` eq/ne; the
/// ordered four map directly.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum CmpOp {
    Eq,
    Ne,
    Lt,
    Le,
    Gt,
    Ge,
}

/// The aggregate functions the subset supports. `Count` doubles for both
/// `COUNT(expr)` and `COUNT(*)` - the latter carries no operand.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum AggFunc {
    Sum,
    Avg,
    Min,
    Max,
    Count,
}

/// A scalar / boolean expression. The same tree covers a projection scalar, a
/// `WHERE` predicate, and an aggregate operand; the lowerer enforces which
/// shapes are legal where (e.g. an aggregate may not nest inside another
/// aggregate). Comparison + boolean nodes only ever appear inside a predicate.
#[derive(Clone, Debug, PartialEq)]
pub enum SqlExpr {
    /// A column reference (`price`).
    Column(String),
    /// A literal scalar.
    Literal(SqlLiteral),
    /// Binary arithmetic over two scalar sub-expressions.
    Arith(ArithOp, Box<SqlExpr>, Box<SqlExpr>),
    /// A comparison; yields a boolean. Legal only inside a predicate.
    Compare(CmpOp, Box<SqlExpr>, Box<SqlExpr>),
    /// Boolean `AND` of two predicates.
    And(Box<SqlExpr>, Box<SqlExpr>),
    /// Boolean `OR` of two predicates.
    Or(Box<SqlExpr>, Box<SqlExpr>),
    /// Boolean `NOT` of a predicate.
    Not(Box<SqlExpr>),
    /// `CASE WHEN <pred> THEN <e> ELSE <e> END`. Lowers to `TsExpr::When`.
    Case {
        when: Box<SqlExpr>,
        then: Box<SqlExpr>,
        otherwise: Box<SqlExpr>,
    },
    /// An aggregate call. `arg` is `None` for `COUNT(*)`.
    Aggregate {
        func: AggFunc,
        arg: Option<Box<SqlExpr>>,
    },
}

impl SqlExpr {
    /// Does this expression contain an aggregate call anywhere in its tree?
    /// Drives the grouped-vs-rowwise lowering decision: a projection list with
    /// any aggregate is an aggregate query.
    pub fn contains_aggregate(&self) -> bool {
        match self {
            SqlExpr::Aggregate { .. } => true,
            SqlExpr::Column(_) | SqlExpr::Literal(_) => false,
            SqlExpr::Arith(_, a, b)
            | SqlExpr::Compare(_, a, b)
            | SqlExpr::And(a, b)
            | SqlExpr::Or(a, b) => a.contains_aggregate() || b.contains_aggregate(),
            SqlExpr::Not(e) => e.contains_aggregate(),
            SqlExpr::Case {
                when,
                then,
                otherwise,
            } => {
                when.contains_aggregate()
                    || then.contains_aggregate()
                    || otherwise.contains_aggregate()
            }
        }
    }
}

/// One item in the `SELECT` projection list.
#[derive(Clone, Debug, PartialEq)]
pub enum SelectItem {
    /// `*` - every source column, in source order.
    Star,
    /// A projected expression with its output column name. The alias is the
    /// `AS` name when present, else a derived name (the bare column name, or a
    /// synthesised `col_N` for a computed expression).
    Expr { expr: SqlExpr, alias: String },
}

/// One `ORDER BY` key.
#[derive(Clone, Debug, PartialEq)]
pub struct OrderKey {
    pub column: String,
    pub ascending: bool,
}

/// A parsed `SELECT` statement. The parser fills every clause it sees; absent
/// clauses are empty / `None`. The lowerer reads this whole struct.
#[derive(Clone, Debug, PartialEq)]
pub struct TsSqlStmt {
    pub projection: Vec<SelectItem>,
    pub table: String,
    pub filter: Option<SqlExpr>,
    pub group_by: Vec<String>,
    pub order_by: Vec<OrderKey>,
    pub limit: Option<usize>,
}
